meetings/2021-04-21.md

```diff
@@ -0,0 +1,63 @@
+# 2021-04-21 Solid Authorization
+
+https://meet.jit.si/solid-authorization
+
+## Agenda
+
+* Minutes https://github.com/solid/authorization-panel/pull/202
+* Minutes https://github.com/solid/authorization-panel/pull/204
+* https://github.com/solid/authorization-panel/pull/201
+
+## Present
+
+* Justin B
+* Henry Story
+* Matthieu Bosquet
+
+## Minutes
+
+[PR 201](https://github.com/solid/authorization-panel/pull/201)
+
+Henry: The PR can be [viewed in full here](ps://github.com/solid/authorization-panel/blob/acp-diff/proposals/wac-acp-diff-story.md).
+
+
+### ACP all of, any of and none of
+
+```typescript
+/**
+ * # The apply policy algorithm
+ *
+ * A policy is applied if:
+ * - At least one matcher it uses matches the given context; and
+ * - All the conditions it defines are satisfied.
+ *
+ * Policy conditions behave like intersection, union and exclusion operators. In
+ * order to be satisfied:
+ * - allOf requires all of its matchers to match the given context;
+ * - anyOf requires one of its matchers to match the given context;
+ * - noneOf requires none of its matchers to match the given context.
+ *
+ * Note: Given that the noneOf condition excludes matches, a policy without a
+ * satisfied allOf or anyOf condition never applies.
+ *
+ * @param policy The policy to evaluate
+ * @param context The context in which the policy is evaluated
+ * @returns True if the policy applies, false otherwise
+ */
+export function applyPolicy(policy: IPolicy, context: IContext): boolean {
+  const matchContext = (matcher: IMatcher) => {
+    return matcher.match(context);
+  };
+
+  // A policy without positive match doesn't apply
+  if (policy.allOf.length + policy.anyOf.length === 0) {
+    return false;
+  }
+
+  return (
+    (policy.allOf.length === 0 || policy.allOf.every(matchContext)) &&
+    (policy.anyOf.length === 0 || policy.anyOf.some(matchContext)) &&
+    (policy.noneOf.length === 0 || !policy.noneOf.some(matchContext))
+  );
+}
+```
```

Comment:
```suggestion
```typescript
/**
 * # The apply policy algorithm
 *
 * A policy is applied if:
 * - At least one matcher it uses matches the given context; and
 * - All the conditions it defines are satisfied.
 *
 * Policy conditions behave like intersection, union and exclusion operators. In
 * order to be satisfied:
 * - allOf requires all of its matchers to match the given context;
 * - anyOf requires one of its matchers to match the given context;
 * - noneOf requires none of its matchers to match the given context.
 *
 * Note: Given that the noneOf condition excludes matches, a policy without a
 * satisfied allOf or anyOf condition never applies.
 *
 * @param policy The policy to evaluate
 * @param context The context in which the policy is evaluated
 * @returns True if the policy applies, false otherwise
 */
export function applyPolicy(policy: IPolicy, context: IContext): boolean {
  const matchContext = (matcher: IMatcher) => {
    return matcher.match(context);
  };

  // A policy without positive match doesn't apply
  if (policy.allOf.length + policy.anyOf.length === 0) {
    return false;
  }

  return (
    (policy.allOf.length === 0 || policy.allOf.every(matchContext)) &&
    (policy.anyOf.length === 0 || policy.anyOf.some(matchContext)) &&
    (policy.noneOf.length === 0 || !policy.noneOf.some(matchContext))
  );
}
```
We started discussing the ACP algorithm for applying policies.
```